pkg/kubelet/pod_workers.go

```diff
@@ -299,13 +319,82 @@ func (s *podSyncStatus) IsTerminated() bool           { return !s.terminatedAt.I
 func (s *podSyncStatus) IsFinished() bool             { return s.finished }
 func (s *podSyncStatus) IsEvicted() bool              { return s.evicted }
 func (s *podSyncStatus) IsDeleted() bool              { return s.deleted }
+func (s *podSyncStatus) IsStarted() bool              { return !s.startedAt.IsZero() }
+
+// WorkType returns this pods' current state of the pod in pod lifecycle state machine.
+func (s *podSyncStatus) WorkType() PodWorkType {
+	if s.IsTerminated() {
+		return TerminatedPodWork
+	}
+	if s.IsTerminationRequested() {
+		return TerminatingPodWork
+	}
+	return SyncPodWork
+}
+
+// replayLastUpdate creates a pendingUpdate for this status if none is currently requested
+// that matches the last update sent. If an update cannot be replayed because we do not have
+// a last update, this method is a no-op and it is the caller's responsibility to ensure
+// the worker goroutine exits gracefully.
+func (s *podSyncStatus) replayLastUpdate() {
+	if s.pendingUpdate != nil || s.lastUpdate == nil {
+		return
+	}
+	copied := *s.lastUpdate
+	s.pendingUpdate = &copied
+}
+
+// mergeLastUpdate records the most recent state from a new update. Pod and MirrorPod are
+// incremented. KillPodOptions is accumulated. If RunningPod is set, Pod is synthetic and
+// will *not* be used as the last pod state unless no previous pod state exists (because
+// the pod worker may be responsible for terminating a pod from a previous run of the
+// kubelet where no config state is visible). The contents of lastUpdate are used as the
+// source of truth for components downstream of the pod workers.
+func (s *podSyncStatus) mergeLastUpdate(other UpdatePodOptions) {
+	opts := s.lastUpdate
+	if opts == nil {
+		opts = &UpdatePodOptions{}
+		s.lastUpdate = opts
+	}
+	if other.Pod != nil {
+		// TODO: Decide whether we need to track whether we have ever received a "full"
+		// pod (i.e. not one created by RunningPod.ToAPIPod()) to avoid giving other
+		// Kubelet components that ask a potentially unfilled pod.
+		if other.RunningPod == nil || opts.Pod == nil {
```

Comment:
There is no current way to reverse predictably ToAPIPod.  But I think in general the bug here is pretending that RunningPod can be a real Pod.  The contract of UpdatePod is clear (RunningPod or Pod, but not both).  It's pretty ugly that we create the fake API pod - most of the callers don't actually need it, so we could instead have UpdatePodOptions have GetUID or something for the key methods.  It would be bad if the synthetic pod were ever used for any other purpose.  Looking at KillPod, we use the synthetic pod for a few global fields like termination policy (wierd) then loop on running pod and call killContainer, which calls preStop hooks (all would be missing).

It's definitely possible that we could get a RunningPod, THEN a real version of the Pod - kubelet is running static pod, then kubelet stopped, static pod config removed, kubelet started, kubelet observes running pod and puts it on lastUpdate, then someone puts the static pod config (with static uid) back in, but if that happens it's indistinguishable from any other static pod update.  So I'll make sure we have a unit test for that case.

I'll review and see whether we could simply have getters on the type to find any common elements, although I suspect it's more that we we want to ensure entry to any other phase except terminating/terminated is impossible for a running pod.  I'll also re-review the merge logic.

Also, in looking at the code, killContainer is doing some sketchy things with grace period, so I'll cue that up as a follow on (roughly, if grace period is zero we don't run the preStop hook, which should apply to force deleted pods, but should also apply to evicted pods, which it doesn't look like it does).

Reply:
Sounds like the cleanest approach is if we can avoid using `ToAPIPod()` in first place and handle the places that need to behave properly with a runtime pod (which sounds like it will only be termination because we won't keep a pod without config around. Maybe we can have a special termination path for runtime pods then (or alternatively killPod could handle a runtime pod OR real api pod, instead of a synthetic API pod).

Regarding killContainer, I think I see that the logic you are referring to:

https://github.com/kubernetes/kubernetes/blob/f9a3fd2810ed4cba159b8d106ca7362c0b528c3a/pkg/kubelet/kuberuntime/kuberuntime_container.go#L697-L700

But it looks like below we always give a 2 sec, `minimumGracePeriodInSeconds` right after... Are you suggesting we also use the 2 seconds for preStop hooks (in case of a force deleted/evicted pod?)

Reply:
> Sounds like the cleanest approach is if we can avoid using ToAPIPod() in first place and handle the places that need to behave properly with a runtime pod (which sounds like it will only be termination because we won't keep a pod without config around)

This is what I went with in current commit, but a special termination path for runtime pods is kind of what we have now.  I'm tempted to verify we can call the normal termination path with a runtime pod (perhaps via an e2e) but that falls into a broader bucket of challenge.

Re: grace period, the grace period isn't an input from the caller, but currently it's not possible to pass down that "we don't want any hooks to run" (eviction probably).  Need to check but I suspect if you have a preStop hook you can hang eviction for a longer time.

Reply:
> This is what I went with in current commit, but a special termination path for runtime pods is kind of what we have now. I'm tempted to verify we can call the normal termination path with a runtime pod (perhaps via an e2e) but that falls into a broader bucket of challenge.

We can try this as followup, i.e. maybe can remove the logic for runtime pod termination and have only a single path as you mentioned. We can give a try and see what blows up in an e2e, and based on that go in and ensure that single path works for both cases.

Reply:
Created a separate method for syncTerminatingRuntimePod